### ISSUE TITLE
libero finetuning

### ISSUE THREAD

**Issue by zhanjiqing** [X-Square-Robot/wall-x#111]
@StarrickLiu @Lum0s517 
Hi, thank you for releasing this great project.

I am trying to fine-tune Wall-X on the LIBERO dataset and then evaluate it with the provided LIBERO evaluation pipeline. I have a few questions about the recommended training setup and expected behavior.

# Config
And my config is: 

model_type: qwen2_5

task:
  dof_config:
    master_right_ee_cartesian_pos: 3     # delta position
    master_right_ee_rotation: 3          # delta rotation (ZYX euler)
    master_right_gripper: 1
    action_padding: 19
  ar_dof_config:
    master_right_ee_cartesian_pos: 3
    master_right_ee_rotation: 3
    master_right_gripper: 1
    action_padding: 19
  agent_pos_config:
    follow_right_ee_cartesian_pos: 3
    follow_right_ee_rotation: 3
    follow_right_gripper: 2
    action_padding: 18
  action_horizon: 10
  action_horizon_flow: 10
  use_state_string_representation: false

model:
  backbone: qwen2_5
  config_path: /home/remote/models/X-Square-Robot--wall-oss-0.5/snapshots/master/config.json
  processor_path: /home/remote/models/Qwen--Qwen2.5-VL-3B-Instruct/snapshots/master
  pretrained_path: /home/remote/models/Qwen--Qwen2.5-VL-3B-Instruct/snapshots/master
  attn_deterministic: true
  use_ema: false
  flow_loss_weight: 1.0
  ar_loss_weight: 0.01

hyperparams:
  num_epoch: 100
  batch_size_per_gpu: 1 #4
  gradient_accumulation_steps: 4
  seed: 10222
  optimizer:
    optimizer_type: dmuon #adamw
    learning_rate: 5.0e-05
    max_grad_norm: 1.0
    enable_grad_clip: true
    betas: [0.9, 0.95]
    weight_decay: 1.0e-8
    eps: 1.0e-8
  scheduler:
    scheduler_type: cosine
    num_warmup_steps: 1000
    num_training_steps: 200000
    min_lr: 1.0e-6

distributed:
  use_fsdp: true
  use_mixed_precision: true
  bf16: true
  sdp_cpu_offload: true
  fsdp_use_orig_params: true
  fsdp_limit_all_gathers: true
  fsdp_forward_prefetch: false

  fsdp_reduce_dtype: bf16

  use_gradient_checkpointing: true
  use_selective_recompute: true

data:
  dataset_type: lerobot
  lerobot_config:
    repo_id: /data/home/share/datasets/HuggingFaceVLA_v21_libero 
    root: null 
  key_mappings:
    camera:
      observation.images.image: face_view
      observation.images.image2: right_wrist_view
    state: observation.state
    action: action
  norm_stats_path: /home/remote/wall-x/workspace/libero20260704/libero/libero_all_norm_stats.json
  train_test_split: 0.95
  num_workers: 4
  max_length: 1024
  resolution:
    face_view: 256
    right_wrist_view: 256

logging:
  log_name: libero_ft
  log_project: lerobot_libero_ft
  log_entity: your_wandb_entity
  use_wandb: false
  log_interval: 10
  save_interval: 2000
  val_interval: 1000000
  epoch_save_interval: 1

checkpoint:
  save_path: /data/home/share/output_wallx
  
  resume_from: /home/remote/models/X-Square-Robot--wall-oss-0.5/snapshots/master/model.safetensors

debug:
  profile: false
  nvtx: false

I am training on 4 RTX 5090 GPUs, each with 32GB VRAM.

# Current observation
I evaluated an early checkpoint trained for less than 3 epochs. The LIBERO evaluation runs successfully without crashing, but the rollout results are all failures, for example:
{"event": "completed", "ep_id": ["libero_spatial", 5, 8], "success": false, "steps": 220}
Most episodes terminate at the maximum number of steps, e.g. steps=220, which I assume corresponds to:
max_infer_times=22 * action_horizon=10

# Questions
## Expected training length
For LIBERO fine-tuning, is the default num_epoch: 100 in workspace/example/libero.yml the recommended training length?
If a checkpoint is trained for only 2-3 epochs on the full LIBERO dataset, is it expected that the LIBERO evaluation success rate is still close to 0?

## Expected learning curve
Roughly at what point should we expect to see non-zero success rates on LIBERO evaluation?
For example, should we expect meaningful success after:
5-10 epochs?
30-50 epochs?
close to 100 epochs?
I understand this depends on batch size and hardware, but any reference numbers would be very helpful.

## DMuon optimizer learning rate
Due to GPU memory constraints, I tried using optimizer_type: dmuon. During training, I see logs like:
lr_group/default/muon: 0.016080
lr_group/default/adamw: 0.000804
lr: 0.016080
This seems much larger than the AdamW learning rate in the default LIBERO config:
optimizer:
  optimizer_type: adamw
  learning_rate: 5.0e-05
Is this large Muon learning rate expected when using DMuon?
For LIBERO fine-tuning, do you recommend using DMuon? If yes, what muon_lr and adamw_lr values would you suggest?

## AdamW vs DMuon for LIBERO
Is the published/recommended LIBERO result obtained with AdamW, DMuon, or another optimizer?
If AdamW is the recommended setup, do you have any suggestions for running it on 4x32GB GPUs, such as:

- gradient checkpointing

- CPU offload

- smaller batch size

- different FSDP settings

Any advice on the recommended LIBERO fine-tuning and evaluation setup would be very helpful.
Thanks!

**Comment by StarrickLiu** [comment]
Hi, apologies for the late reply — thanks for the detailed report and configs, that made this easy to diagnose.

Three things are causing the zero success rate, and all are fixable:

**1. Global batch size — use 128.**
Your current setup is `batch_size_per_gpu: 1` × `gradient_accumulation_steps: 4` × 4 GPUs = a global batch of **16**, which is far too small for stable fine-tuning. We recommend a **global batch size of 128**. On your 4×32GB cards, reach it with e.g.:

```yaml
hyperparams:
  batch_size_per_gpu: 1
  gradient_accumulation_steps: 32   # 4 GPUs × 1 × 32 = 128
```

(increase `batch_size_per_gpu` and lower `gradient_accumulation_steps` proportionally if memory allows).

**2. Use DMuon — on LIBERO it works significantly better than plain AdamW.**
In our experiments DMuon converges much better than pure AdamW for LIBERO fine-tuning, so we recommend it (it's also friendlier on memory, which helps on 32GB cards).

**3. Learning rates — this is the main issue in your run.**
The values you saw (`muon: 0.016080`, `adamw: 0.000804`) are far too high. There's an important gotcha here: **DMuon reads `muon_lr` and `adamw_lr`, not `learning_rate`.** So setting `learning_rate: 0.0001` has no effect under DMuon, and it silently falls back to the defaults (`muon_lr=0.02`, `adamw_lr=0.001`) — which are ~200×/20× larger than what LIBERO needs, so training diverges.

Please set them explicitly to our recommended values:

```yaml
optimizer:
  optimizer_type: dmuon
  muon_lr: 0.0001      # 1e-4
  adamw_lr: 0.00005    # 5e-5
```

**What a healthy run should look like.** With the correct global batch size and these LRs, you should already see a **success rate of ~50% or higher after the very first fine-tuning epoch**. If you don't, there's no point training longer — stop and investigate the cause first.

The most important thing to watch is the **loss curve on your dashboard**. A healthy run converges smoothly. If you see a **loss spike or otherwise abnormal convergence**, that's a strong signal the learning rate is too high (or a similar problem) — stop, check, and fix it rather than training through it. That's almost certainly what happened here with the too-large default LRs.

For FSDP on 4×32GB, keep FSDP + gradient checkpointing enabled; DMuon's sharding further reduces the optimizer-state footprint.

We'll also update the example configs so the recommended `muon_lr`/`adamw_lr` are set explicitly (rather than relying on the too-large defaults). Thanks again for surfacing this!
